### PR TITLE
Added connection pooling to DB config and added defaults for the same

### PR DESCRIPTION
**Pull request by pmahindrakar-oss**
Signed-off-by: Prafulla Mahindrakar <prafulla.mahindrakar@gmail.com>

# TL;DR
Added connection pooling for the admin DB config.
Conservative defaults have been added based on the RDS limits.

## Type
 - [X] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [X] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/1956

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/runtime/interfaces/application_configuration.go [hunk 1]
```diff
@@ -18,6 +21,9 @@ type DbConfig struct {
 	DeprecatedDebug        bool   `json:"debug" pflag:",deprecated"`
 
 	EnableForeignKeyConstraintWhenMigrating bool           `json:"enableForeignKeyConstraintWhenMigrating" pflag:",Whether to enable gorm foreign keys when migrating the db"`
+	MaxIdleConnections                      int            `json:"maxIdleConnections" pflag:",maxIdleConns sets the maximum number of connections in the idle connection pool."`
+	MaxOpenConnections                      int            `json:"maxOpenConnections" pflag:",maxOpenConns sets the maximum number of open connections to the database."`
+	ConnMaxLifeTime                         time.Duration  `json:"connMaxLifeTime" pflag:",sets the maximum amount of time a connection may be reused"`
```
Comment (EngHabu) [inline pkg/runtime/interfaces/application_configuration.go:26]: I think you have to use config.Duration to get the right unmarshaling behavior from commandline...

```suggestion
	ConnMaxLifeTime                         config.Duration  `json:"connMaxLifeTime" pflag:",sets the maximum amount of time a connection may be reused"`
```

**Review by EngHabu — APPROVED** [review]
Awesome! one comment

File: pkg/runtime/interfaces/application_configuration.go [hunk 2]
```diff
@@ -17,8 +18,11 @@ type DbConfig struct {
 	DeprecatedExtraOptions string `json:"options" pflag:",deprecated"`
 	DeprecatedDebug        bool   `json:"debug" pflag:",deprecated"`
 
-	EnableForeignKeyConstraintWhenMigrating bool           `json:"enableForeignKeyConstraintWhenMigrating" pflag:",Whether to enable gorm foreign keys when migrating the db"`
-	PostgresConfig                          PostgresConfig `json:"postgres"`
+	EnableForeignKeyConstraintWhenMigrating bool            `json:"enableForeignKeyConstraintWhenMigrating" pflag:",Whether to enable gorm foreign keys when migrating the db"`
+	MaxIdleConnections                      int             `json:"maxIdleConnections" pflag:",maxIdleConns sets the maximum number of connections in the idle connection pool."`
+	MaxOpenConnections                      int             `json:"maxOpenConnections" pflag:",maxOpenConns sets the maximum number of open connections to the database."`
```
Comment (lupasarin) [inline pkg/runtime/interfaces/application_configuration.go:23]: I would use the full name in the description to avoid confusion, i.e. maxIdleConnections instead of maxIdleConns and maxOpenConnections instead of maxOpenConns. For example: "maxIdleConnections sets the maximum number of connections in the idle connection pool."

**Review by lupasarin — DISMISSED** [review]

File: pkg/runtime/interfaces/application_configuration.go (hunk 2 above)
Comment (pmahindrakar-oss) [reply pkg/runtime/interfaces/application_configuration.go]: Done

File: pkg/repositories/database_test.go [hunk 3]
```diff
@@ -61,3 +67,50 @@ func TestGetPostgresDsn(t *testing.T) {
...
+			DeprecatedPort:     5432,
+			MaxIdleConnections: 10,
+			MaxOpenConnections: 1000,
+			ConnMaxLifeTime:    config.Duration{Duration: time.Hour},
+		}
+		err = setupDbConnectionPool(gormDb, dbConfig)
+		assert.Nil(t, err)
+		genericDb, err := gormDb.DB()
+		assert.Nil(t, err)
+		assert.Equal(t, genericDb.Stats().MaxOpenConnections, 1000)
```
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply pkg/repositories/database_test.go:85]: Other fields are currently not exposed in the stats object and there is no other function in the sql.DB that exposes these fields. Hence only added for MaxOpenConnections

File: pkg/repositories/database_test.go [hunk 4]
```diff
@@ -61,3 +67,50 @@ func TestGetPostgresDsn(t *testing.T) {
...
+		dbConfig := &runtimeInterfaces.DbConfig{
+			DeprecatedPort:     5432,
+			MaxIdleConnections: 10,
+			MaxOpenConnections: 1000,
+		}
+		err = setupDbConnectionPool(gormDb, dbConfig)
+		assert.Nil(t, err)
+		genericDb, err := gormDb.DB()
+		assert.Nil(t, err)
+		assert.Equal(t, genericDb.Stats().MaxOpenConnections, 1000)
```
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply pkg/repositories/database_test.go:99]: Same as above

**Review by EngHabu — APPROVED** [review]
